promql/engine.go

```diff
@@ -2903,6 +2903,263 @@ func scalarBinop(op parser.ItemType, lhs, rhs float64) float64 {
+				removedSum += bucket.Count * bucketMidpoint
+				trimmedHist.NegativeBuckets[i] = 0
+			}
+		}
+
+		// For TRIM_LOWER, we keep observations above the trim point (rhs)
+	case parser.TRIM_LOWER:
+		for i, iter := 0, trimmedHist.PositiveBucketIterator(); iter.Next(); i++ {
+			hasPositive = true
+			bucket := iter.At()
+			keepCount, bucketMidpoint := computeBucketTrim(op, bucket, rhs, true, isCustomBucket)
+
+			switch {
+			case bucket.Lower >= rhs:
+				updatedCount += bucket.Count
+			case bucket.Upper > rhs:
+				removedCount := bucket.Count - keepCount
+				removedMid := bucketMidpoint
+				removedSum += removedCount * removedMid
+
+				trimmedHist.PositiveBuckets[i] = keepCount
+				updatedCount += keepCount
+			default:
+				if !isCustomBucket {
+					bucketMidpoint = math.Sqrt(bucket.Lower * bucket.Upper)
+				}
+				removedSum += bucket.Count * bucketMidpoint
+				trimmedHist.PositiveBuckets[i] = 0
+			}
+		}
+
+		for i, iter := 0, trimmedHist.NegativeBucketIterator(); iter.Next(); i++ {
+			hasNegative = true
+			bucket := iter.At()
+			keepCount, bucketMidpoint := computeBucketTrim(op, bucket, rhs, false, isCustomBucket)
+
+			switch {
+			case bucket.Lower >= rhs:
+				updatedCount += bucket.Count
+			case bucket.Upper > rhs:
+				removedCount := bucket.Count - keepCount
+				removedMid := bucketMidpoint
+				removedSum += removedCount * removedMid
+
+				trimmedHist.NegativeBuckets[i] = keepCount
+				updatedCount += keepCount
+			default:
+				bucketMidpoint = math.Sqrt(bucket.Lower * bucket.Upper)
+				removedSum += bucket.Count * bucketMidpoint
+				trimmedHist.NegativeBuckets[i] = 0
+			}
+		}
+	}
+
+	// Handle the zero count bucket
+	if trimmedHist.ZeroCount > 0 {
+		zeroBucket := trimmedHist.ZeroBucket()
+		zLower := zeroBucket.Lower
+		zUpper := zeroBucket.Upper
+
+		switch op {
+		case parser.TRIM_UPPER:
+			switch {
+			case rhs < zLower:
+				trimmedHist.ZeroCount = 0
+			case rhs > zUpper:
+				updatedCount += trimmedHist.ZeroCount
+			default:
+				fraction := (rhs - zLower) / (zUpper - zLower)
+				keepCount := trimmedHist.ZeroCount * fraction
+				trimmedHist.ZeroCount = keepCount
+				updatedCount += keepCount
+			}
+
+		case parser.TRIM_LOWER:
+			switch {
+			case rhs > zUpper:
+				trimmedHist.ZeroCount = 0
+			case rhs < zLower:
+				updatedCount += trimmedHist.ZeroCount
+			default:
+				fraction := (zUpper - rhs) / (zUpper - zLower)
+				keepCount := trimmedHist.ZeroCount * fraction
+				trimmedHist.ZeroCount = keepCount
+				updatedCount += keepCount
+			}
+		}
+	}
+
+	// Apply new sum
+	newSum := origSum - removedSum
+
+	// Clamp correction
+	if !hasNegative && newSum < 0 {
+		newSum = 0
+	}
+	if !hasPositive && newSum > 0 {
+		newSum = 0
+	}
+
+	// Update the histogram's count and sum
+	trimmedHist.Count = updatedCount
+	trimmedHist.Sum = newSum
+
+	trimmedHist.Compact(0)
+}
+
+func computeMidpoint(b histogram.Bucket[float64], product float64, isCustom, isPositive bool) float64 {
```

Comment:
Following from my comment above I would expect this function to simply look like this:
```
func computeMidpoint(a, b float64, isLinear, isPositive bool) float64 {
	if isLinear {
		return (b + a) / 2
	}

	if isPositive {
		return math.Sqrt(math.Abs(a) * math.Abs(b))
	}
	return -math.Sqrt(math.Abs(a) * math.Abs(b))
}
```

But I could not get every test to pass. So I think that question needs answering about why we use different kind of ranges when calling this function. Bucket lower/upper bound for custom buckets, but range adjusted with `rhs` for the exponential buckets.